### PR TITLE
Add Nutanix-specific platform types

### PR DESCRIPTION
**Pull request by thunderboltsid**
Add types, defaults, and validations to support Nutanix platform on the Openshift installer.

### CODE REVIEW

File: pkg/types/nutanix/OWNERS [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# See the OWNERS docs: https://git.k8s.io/community/contributors/guide/owners.md
+# This file just uses aliases defined in OWNERS_ALIASES.
+
+approvers:
+  - nutanix-approvers
```
Comment (staebler) [inline pkg/types/nutanix/OWNERS:5]: Add a reviewers section, too.

File: pkg/types/nutanix/machinepool.go [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+	// NumCoresPerSocket is the number of cores per socket in a vm. The number
+	// of vCPUs on the vm will be NumCPUs/NumCoresPerSocket.
+	//
+	// +optional
+	NumCoresPerSocket int32 `json:"coresPerSocket"`
+
+	// Memory is the size of a VM's memory in MB.
+	//
+	// +optional
+	MemoryMiB int64 `json:"memoryMB"`
```
Comment (staebler) [inline pkg/types/nutanix/machinepool.go:20]: Is this MiB or MB? It is not consistent.

File: pkg/types/nutanix/machinepool.go [hunk 3]
```diff
@@ -0,0 +1,57 @@
+package nutanix
+
+// MachinePool stores the configuration for a machine pool installed
+// on Nutanix.
+type MachinePool struct {
+	// NumCPUs is the total number of virtual processor cores to assign a vm.
+	//
+	// +optional
+	NumCPUs int32 `json:"cpus"`
```
Comment (staebler) [inline pkg/types/nutanix/machinepool.go]: Optionals should have `omitempty`.
```suggestion
	NumCPUs int32 `json:"cpus,omitempty"`
```

File: pkg/types/nutanix/metadata.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+	// PrismCentral is the domain name or IP address of the Prism Central.
+	PrismCentral string `json:"prismCentral"`
+	// Username is the name of the user to use to connect to the Prism Central.
+	Username string `json:"username"`
+	// Password is the password for the user to use to connect to the Prism Central.
+	Password string `json:"password"`
+	// Port is the port used to connect to the Prism Central.
+	Port string `json:"port"`
+	// Disable certificate checking when connecting to Prism Central.
+	Insecure bool `json:"insecure"`
```
Comment (staebler) [inline pkg/types/nutanix/metadata.go:14]: We should not be supporting insecure connections. Let's not repeat the mistake made for vSphere.

File: pkg/types/nutanix/platform.go [hunk 5]
```diff
@@ -0,0 +1,49 @@
+package nutanix
+
+// Platform stores any global configuration used for Nutanix platforms.
+type Platform struct {
+	// PrismCentral is the domain name or IP address of the Prism Central.
+	PrismCentral string `json:"prismCentral"`
+
+	// Insecure disables certificate checking when connecting to Prism Central.
+	Insecure bool `json:"insecure"`
```
Comment (staebler) [inline pkg/types/nutanix/platform.go:9]: No insecure connections.

File: pkg/types/nutanix/platform.go [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+	Username string `json:"username"`
+
+	// Password is the password for the user to use to connect to the Prism Central.
+	Password string `json:"password"`
+
+	// PrismElement is the name of the Prism Element cluster to use in the Prism Central.
+	PrismElement string `json:"prismElement"`
+
+	// DefaultStorageContainer is the default datastore to use for provisioning volumes.
+	// DefaultStorageContainer string `json:"defaultStorageContainer"`
```
Comment (staebler) [inline pkg/types/nutanix/platform.go:24]: Remove these rather than merging in commented out code.

File: pkg/types/nutanix/platform.go [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+	APIVIP string `json:"apiVIP,omitempty"`
+
+	// IngressVIP is the virtual IP address for ingress
+	//
+	// +kubebuilder:validation:format=ip
+	// +optional
+	IngressVIP string `json:"ingressVIP,omitempty"`
+
+	// DefaultMachinePlatform is the default configuration used when
+	// installing on VSphere for machine pools which do not define their own
```
Comment (staebler) [inline pkg/types/nutanix/platform.go]:
```suggestion
	// installing on Nutanix for machine pools which do not define their own
```

File: pkg/types/nutanix/platform.go [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+	// platform configuration.
+	// +optional
+	DefaultMachinePlatform *MachinePool `json:"defaultMachinePlatform,omitempty"`
+
+	// Network specifies the name of the network to be used by the cluster.
```
Comment (staebler) [inline pkg/types/nutanix/platform.go:47]:
```suggestion
	// SubnetUUID specifies the name of the network to be used by the cluster.
```

File: pkg/types/defaults/installconfig.go [hunk 9]
```diff
@@ -107,5 +108,8 @@ func SetInstallConfigDefaults(c *types.InstallConfig) {
 		}
 	case c.Platform.None != nil:
 		nonedefaults.SetPlatformDefaults(c.Platform.None)
+	case c.Platform.Nutanix != nil:
+		nutanixdefaults.SetPlatformDefaults(c.Platform.Nutanix, c)
```
Comment (staebler) [inline pkg/types/defaults/installconfig.go:112]: The install config is not used. Do not pass it in.
```suggestion
		nutanixdefaults.SetPlatformDefaults(c.Platform.Nutanix)
```

File: pkg/types/nutanix/validation/platform.go [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+		allErrs = append(allErrs, field.Required(fldPath.Child("subnet"), "must specify the subnet"))
+	}
+	if len(p.PrismCentral) != 0 {
+		if err := validate.Host(p.PrismCentral); err != nil {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("prismCentral"), p.PrismCentral, "must be the domain name or IP address of the Prism Central"))
+		}
+	}
+
+	// If all VIPs are empty, skip IP validation.  All VIPs are required to be defined together.
+	if strings.Join([]string{p.APIVIP, p.IngressVIP}, "") != "" {
```
Comment (staebler) [inline pkg/types/nutanix/validation/platform.go:41]: This cleverness reduces readability.
```suggestion
	if p.APIVIP != "" || p.IngressVIP != "" {
```

File: pkg/types/nutanix/validation/platform.go [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+		allErrs = append(allErrs, validateVIPs(p, fldPath)...)
+	}
+
+	return allErrs
+}
+
+// ValidateForProvisioning checks that the specified platform is valid.
+func ValidateForProvisioning(p *nutanix.Platform, fldPath *field.Path) field.ErrorList {
```
Comment (staebler) [inline pkg/types/nutanix/validation/platform.go:49]: This function should be validating that resources that need to exist in the Nutanix provider exist and are configured as expected. This function is not a place to do static validation checking.

File: pkg/types/nutanix/validation/platform.go [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("apiVIP"), p.APIVIP, err.Error()))
+	}
+
+	if len(p.IngressVIP) == 0 {
+		allErrs = append(allErrs, field.Required(fldPath.Child("ingressVIP"), "must specify a VIP for Ingress"))
+	} else if err := validate.IP(p.IngressVIP); err != nil {
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("ingressVIP"), p.IngressVIP, err.Error()))
+	}
+
+	if len(p.APIVIP) != 0 && len(p.IngressVIP) != 0 && p.APIVIP == p.IngressVIP {
```
Comment (staebler) [inline pkg/types/nutanix/validation/platform.go:80]: At this point in the code, we are already guaranteed that at least one of the VIPs are non-empty.
```suggestion
	if p.APIVIP == p.IngressVIP {
```

File: pkg/types/validation/installconfig.go [hunk 13]
```diff
@@ -500,6 +502,11 @@ func validatePlatform(platform *types.Platform, fldPath *field.Path, network *ty
 			return ovirtvalidation.ValidatePlatform(platform.Ovirt, f)
 		})
 	}
+	if platform.Nutanix != nil {
+		validate(nutanix.Name, platform.Nutanix, func(f *field.Path) field.ErrorList {
+			return nutanixvalidation.ValidatePlatform(platform.Nutanix, f)
```
Comment (staebler) [inline pkg/types/validation/installconfig.go:507]: This should not be doing static validation. This should only be doing validation that requires connection to the Nutanix provider.

File: pkg/types/nutanix/machinepool.go [hunk 14]
```diff
@@ -6,13 +6,13 @@ type MachinePool struct {
 	// NumCPUs is the total number of virtual processor cores to assign a vm.
 	//
 	// +optional
-	NumCPUs int32 `json:"cpus"`
+	NumCPUs int64 `json:"cpus"`
```
Comment (staebler) [inline pkg/types/nutanix/machinepool.go:9]: Squash all the changes from this commit into the previous commit where these fields were originally defined.

File: pkg/types/nutanix/platform.go [hunk 15]
```diff
@@ -39,11 +39,12 @@ type Platform struct {
...
 	// DefaultMachinePlatform is the default configuration used when
-	// installing on VSphere for machine pools which do not define their own
+	// installing on Nutanix for machine pools which do not define their own
 	// platform configuration.
 	// +optional
 	DefaultMachinePlatform *MachinePool `json:"defaultMachinePlatform,omitempty"`
 
 	// Network specifies the name of the network to be used by the cluster.
-	Subnet string `json:"subnet,omitempty"`
+	SubnetUUID string `json:"subnetUuid,omitempty"`
```
Comment (staebler) [inline pkg/types/nutanix/platform.go:48]:
```suggestion
	SubnetUUID string `json:"subnetUUID,omitempty"`
```

File: pkg/types/nutanix/platform.go [hunk 16]
```diff
@@ -18,7 +18,7 @@ type Platform struct {
 	Password string `json:"password"`
 
 	// PrismElement is the name of the Prism Element cluster to use in the Prism Central.
-	PrismElement string `json:"prismElement"`
+	PrismElementUUID string `json:"prismElementUuid"`
```
Comment (staebler) [inline pkg/types/nutanix/platform.go:21]: Squash this commit.

Comment (staebler) [inline pkg/types/nutanix/platform.go:21]:
```suggestion
	PrismElementUUID string `json:"prismElementUUID"`
```

File: go.mod [hunk 17]
```diff
@@ -77,6 +77,7 @@ require (
 	github.com/terraform-providers/terraform-provider-azurestack v0.10.0
 	github.com/terraform-providers/terraform-provider-ignition/v2 v2.1.0
 	github.com/terraform-providers/terraform-provider-local v1.4.0
+	github.com/terraform-providers/terraform-provider-nutanix v1.1.0
```
Comment (staebler) [inline go.mod:80]: Why is this needed for this PR?

File: pkg/types/nutanix/client.go [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+	"fmt"
+	"time"
+
+	nutanixClient "github.com/terraform-providers/terraform-provider-nutanix/client"
+	nutanixClientV3 "github.com/terraform-providers/terraform-provider-nutanix/client/v3"
+	"k8s.io/klog"
+)
+
+
+func CreateNutanixClient(ctx context.Context, prismCentral, port, username, password string, insecure bool) (*nutanixClientV3.Client, error) {
```
Comment (staebler) [inline pkg/types/nutanix/client.go:14]: This does not seem appropriate to include in this PR, given that the scope of this PR is limited to adding the platform types.

**Review by staebler — COMMENTED** [review]
Please separate vendor changes into their own commits, to make reviewing and perusing history easier.

File: go.mod (hunk 17 above)
Comment (thunderboltsid) [reply go.mod:80]: This is used to create a client that is used in later PRs.

File: pkg/types/nutanix/machinepool.go (hunk 14 above)
Comment (thunderboltsid) [reply pkg/types/nutanix/machinepool.go:9]: Okay.

File: pkg/types/nutanix/machinepool.go (hunk 2 above)
Comment (thunderboltsid) [reply pkg/types/nutanix/machinepool.go:20]: Good catch. It is MiB.

File: pkg/types/nutanix/platform.go (hunk 16 above)
Comment (thunderboltsid) [reply pkg/types/nutanix/platform.go:21]: Okay.

File: pkg/types/defaults/installconfig.go (hunk 9 above)
Comment (thunderboltsid) [reply pkg/types/defaults/installconfig.go:112]: ok

File: pkg/types/nutanix/OWNERS (hunk 1 above)
Comment (thunderboltsid) [reply pkg/types/nutanix/OWNERS:5]: ok

File: pkg/types/nutanix/validation/platform.go (hunk 10 above)
Comment (thunderboltsid) [reply pkg/types/nutanix/validation/platform.go:41]: fair enough :)

File: pkg/types/nutanix/validation/platform.go (hunk 11 above)
Comment (thunderboltsid) [reply pkg/types/nutanix/validation/platform.go:47]: Fair point. We will create future PRs that will switch the static validation to resource validation i.e. see if a prism element with the UUID actually exists on a prism central cluster.

File: pkg/types/validation/installconfig.go (hunk 13 above)
Comment (thunderboltsid) [reply pkg/types/validation/installconfig.go:507]: We will address that in a future PR.

File: pkg/types/nutanix/platform.go (hunk 5 above)
Comment (thunderboltsid) [reply pkg/types/nutanix/platform.go:9]: For context. This is still https. We just allow disabling cert checking due to self signed certs.

File: pkg/types/nutanix/metadata.go (hunk 4 above)
Comment (thunderboltsid) [reply pkg/types/nutanix/metadata.go:14]: For context. This is still https. We just allow disabling cert checking due to self signed certs.

Comment (staebler) [reply pkg/types/nutanix/metadata.go:14]: We should not be allowing insecure to support self-signed certs. We should be requiring the user to provide the CA bundle to be able to trust the self-signed certs.

Comment (rvanderp3) [reply pkg/types/nutanix/metadata.go:14]: +1 to staebler's guidance on requiring a CA bundle to trust self-signed certs.  The team is having discussions on how the CA bundle can be obtained.

Comment (vnephologist) [reply pkg/types/nutanix/metadata.go:14]: @thunderboltsid we will just have to doc download/import of certs like here: https://docs.openshift.com/container-platform/4.9/installing/installing_vsphere/installing-vsphere-installer-provisioned.html#installation-adding-vcenter-root-certificates_installing-vsphere-installer-provisioned

File: pkg/types/nutanix/platform.go (hunk 5 above)
Comment (vnephologist) [reply pkg/types/nutanix/platform.go:9]: @thunderboltsid we will just have to doc download/import of certs like here: https://docs.openshift.com/container-platform/4.9/installing/installing_vsphere/installing-vsphere-installer-provisioned.html#installation-adding-vcenter-root-certificates_installing-vsphere-installer-provisioned

File: pkg/types/nutanix/platform.go [hunk 19]
```diff
@@ -0,0 +1,46 @@
+package nutanix
+
+// Platform stores any global configuration used for Nutanix platforms.
+type Platform struct {
```
Comment (rvanderp3) [inline pkg/types/nutanix/platform.go:4]: Enhancement 912 references a [container](https://github.com/openshift/enhancements/pull/912/files#diff-74fc807d5f6f39d7ce057f41e1197572515da4c728dbc70dcfc91668ed7486b9R87) that is required.  Is that still required?

Comment (thunderboltsid) [reply pkg/types/nutanix/platform.go:4]: Good catch. Added it.

File: pkg/types/nutanix/platform.go [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+	// Port is the port to use to connect to the Prism Central.
+	Port string `json:"port"`
+
+	// Username is the name of the user to use to connect to the Prism Central.
+	Username string `json:"username"`
+
+	// Password is the password for the user to use to connect to the Prism Central.
+	Password string `json:"password"`
+
+	// PrismElement is the name of the Prism Element cluster to use in the Prism Central.
```
Comment (staebler) [inline pkg/types/nutanix/platform.go:17]: Is this a name or a UUID?
```suggestion
	// PrismElementUUID is the UUID of the Prism Element cluster to use in the Prism Central.
```

File: pkg/types/nutanix/platform.go [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+	PrismElementUUID string `json:"prismElementUUID"`
+
+	// DefaultStorageContainer is the default datastore to use for provisioning volumes.
+	DefaultStorageContainer string `json:"defaultStorageContainer"`
+
+	// ClusterOSImage overrides the url provided in rhcos.json to download the RHCOS Image
+	ClusterOSImage string `json:"clusterOSImage,omitempty"`
```
Comment (staebler) [inline pkg/types/nutanix/platform.go:24]:
```suggestion
	//
	// +optional
	ClusterOSImage string `json:"clusterOSImage,omitempty"`
```

File: pkg/types/nutanix/platform.go [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+	IngressVIP string `json:"ingressVIP,omitempty"`
+
+	// DefaultMachinePlatform is the default configuration used when
+	// installing on Nutanix for machine pools which do not define their own
+	// platform configuration.
+	// +optional
+	DefaultMachinePlatform *MachinePool `json:"defaultMachinePlatform,omitempty"`
+
+	// SubnetUUID specifies the UUID of the subnet to be used by the cluster.
+	SubnetUUID string `json:"subnetUUID,omitempty"`
```
Comment (staebler) [inline pkg/types/nutanix/platform.go:45]:
```suggestion
	//
	// +optional
	SubnetUUID string `json:"subnetUUID,omitempty"`
```

File: pkg/types/nutanix/validation/platform.go (hunk 11 above)
Comment (staebler) [reply pkg/types/nutanix/validation/platform.go:50]: Remove the static validation, as there is no value in having it in here now. The future PRs can add resource validation, but they do not need to replace static validation.

**Review by staebler — COMMENTED** [review]
Looks good except or a couple remaining nits.

File: pkg/types/validation/installconfig.go (hunk 13 above)
Comment (staebler) [reply pkg/types/validation/installconfig.go:507]: Sorry, my comment here was incorrect. This is right place to do static validation.

**Comment by staebler** [comment]
There are some formatting errors.
```
INFO[2022-02-22T14:09:45Z] diff --git a/pkg/types/validation/installconfig_test.go b/pkg/types/validation/installconfig_test.go
```
And some unit test errors.
```
FAIL	github.com/openshift/installer/pkg/explain	0.083s
```

File: pkg/types/nutanix/platform.go (hunk 20 above)
Comment (thunderboltsid) [reply pkg/types/nutanix/platform.go:17]: Good catch!

File: pkg/types/nutanix/validation/platform.go (hunk 11 above)
Comment (thunderboltsid) [reply pkg/types/nutanix/validation/platform.go:50]: @staebler Looking at the validation package for different cloud providers, most of them are also doing just static validation.

Comment (thunderboltsid) [reply pkg/types/nutanix/validation/platform.go]: Removed `ValidateForProvisioning`

**Review by staebler — COMMENTED** [review]
Just a small nit. Otherwise it looks good to me.
Feel free to re-apply the lgtm when the comment is addressed.
/approve
/lgtm cancel

File: pkg/types/machinepools.go [hunk 23]
```diff
@@ -133,8 +137,12 @@ func (p *MachinePoolPlatform) Name() string {
 		return vsphere.Name
 	case p.Ovirt != nil:
 		return ovirt.Name
+
 	case p.PowerVS != nil:
 		return powervs.Name
+
```
[the remark this answers is not in the document]
Comment (thunderboltsid) [reply pkg/types/machinepools.go]: Good catch. fixed.

**Comment by thunderboltsid** [comment]
> 

@staebler Seems like the lgtm got removed but the approve didn't get added 😅
